docs/proposals/templates.md

```diff
@@ -0,0 +1,349 @@
+envelope containing arbitrary content is converted into a list of k8s objects.
+
+The actual mechanism for looking up and invoking a template expansion plugin (as well as for how one is provided to k8s) is
+still being defined.  One suggestion was to essentially run a template expansion plugin as a microserver on the k8s cluster
+itself, possibly within a namespace with a name that matched the `templateKind`.  However dictating specific namespaces
+be available within the cluster for a plugin feels overly burdensome.  A map between the `templateKind` and
+the api endpoint capable of expanding it is necessary, however.
+
+Template processors could also be provided in client form to help template authors perform local testing+debugging.
+
+Nested templates should be handled by the primary endpoint, retrieving and handing each nested template off to the appropriate
+expansion plugin, and then assembling the aggregate list of objects from the top level template and any nested templates.
+
+## Open discussion points
+### Where to define parameters (closed?)
+
+There has been some discussion around where to define parameters that are being injected into a Template
+
+1. In a separate standalone file
+2. Within the Template itself
+
+This proposal suggests including the parameter definitions within the Template, which provides a self-contained structure that
+can be easily versioned, transported, and instantiated without risk of mismatching content.  In addition, a Template can easily
+be validated to confirm that all parameter references are resolveable.
+
+Separating the parameter definitions makes for a more complex process with respect to
+* Editing a template (if/when first class editing tools are created)
+* Storing/retrieving template objects with a central store
+
+### How to define parameters (closed?)
+
+This question should now be closed as the problem is delegated to individual template expansion plugins.  They will
+be provided with a map of parameter values and metadata and can inject/substitute those parameters into the 
+template body based on the scheme they define.
+
+### Where to store templates
+
+It has been suggested that k8s should enable template storage with the benefits being:
+
+* Allows the user to see and share available templates
+* Provides unified access control
+* Simplified service catalog flows in which instantiating a service is equivalent to instantiating a template
+* Templates can still be stored elsewhere for versioning/revision control, with the currently accepted version
+  residing in the k8s store for use.
+
+However there are also concerns that storing templates doesn't fit with the primary mission of the k8s etcd store,
+which is to store the definitions of objects being actively managed/run by k8s.
+
+### Supporting generated values
+
+There is a use case for being able to define a parameter value that should have an auto-generated value on
```

Comment:
changing "how to define" parameters to "Where to define parameters" instead.  I think this is a separate issue.  "where to define parameters" meant literally where in the file (or in separate file) to define them.